### ISSUE TITLE
Performance degradation between 2.0.10 and 2.0.11

### ISSUE THREAD

**Issue by brettwooldridge** [lukehutch/fast-classpath-scanner#117]
We recently noticed our application startup time jump from ~4.7s to ~6.0s.  We initially traced the performance degration to our upgrade from v2.0.5 to v2.0.21.  However, a further binary search ... trying v2.0.15, v2.0.10, v.2.0.12, etc. lead us to pinpoint the drop in performance as occurring between v2.0.10 and v2.0.11.

Any idea which change was responsible for the regression?

**Comment by lukehutch** [comment]
Interesting, thanks for doing the bisection. Can you please attach a log (calling .verbose() before .scan()) for each of 2.0.10 and 2.0.11?

**Comment by brettwooldridge** [comment]
@lukehutch Here ya go.  Rather large ... ~125MB each ... but zipped down to ~7MB for both.

[verbose-logs.zip](https://github.com/lukehutch/fast-classpath-scanner/files/1026279/verbose-logs.zip)

**Comment by brettwooldridge** [comment]
@lukehutch While I've got your ear, our application is comprised of roughly 85 jars (255 if you count third-party jars), and we only scan our jars ... any tips on cutting scanning times (even off of the v2.0.10 levels)?

**Comment by lukehutch** [comment]
@brettwooldridge the line count discrepancy doesn't surprise me -- I went back and looked at the checkins between these two versions, and the main change relevant to your problem is that 2.0.11 finds more `ClassLoaders` -- so it is probably scanning more jars. Will look into it in some detail. (It's a bit surprising that 255 jars take 4-6 seconds, but maybe you have some big jars...) 

Just curious -- did you try other classpath scanners? I put a lot of work into squeezing every drop of performance out of classpath scanning for FastClasspathScanner, but it is possible that a more specialized scanner (e.g. one that just scans for annotations) could work faster than FastClasspathScanner.

**Comment by brettwooldridge** [comment]
@lukehutch We have not tried other scanners, we've been very happy with FastClasspathScanner.

To be clear, the scanning does *not* take 4-6 seconds.  Our whole application startup takes 4-6 seconds.  But given that a tremendous amount of stuff is happening during that startup, completely unrelated to scanning, the ~1.5 second jump in startup from v2.0.10 to v2.0.11 was surprising.  I haven't measured total scanning time in v2.0.10, but intuitively I feel like it's in the neighborhood of the low hundreds of milliseconds.  v2.0.11+ clearly tacks at least a second onto that.

**Comment by lukehutch** [comment]
@brettwooldridge First thing I noticed: you are scanning a total of *thirteen times* in these log files! Unless you need 13 different sets of scan criteria, you can accomplish all the scanning work in a single pass.

If you look right at the end of each logfile, you will see that scanning took 0.044 seconds with version 2.0.10, and that goes up to 0.214 seconds with version 2.0.11. Multiply that by 13, and you get a 2.21s difference in scan time.

I need to dig into why the scan time increased so much, but the point is that at worst, scanning should now take a total of 0.214 seconds, not that number times 13.

**Comment by lukehutch** [comment]
So the difference between scan times for the two versions does indeed come down to version 2.0.11 finding more jars to scan.
[jars-scanned-by-both.txt](https://github.com/lukehutch/fast-classpath-scanner/files/1026478/jars-scanned-by-both.txt)
[extra-jars-scanned-by-2.0.11.txt](https://github.com/lukehutch/fast-classpath-scanner/files/1026477/extra-jars-scanned-by-2.0.11.txt)
There are some extension jars that are not correctly identified as system jars by 2.0.11 (I'll open a bug for that, although I don't have Mac OS X to test with). But there are also a ton of other project jars that 2.0.11 finds that 2.0.10 does not find. Please take a look at the lists above and let me know what is going on with these extra jars.

**Comment by brettwooldridge** [comment]
@lukehutch That "extra-jars-scanned" list is essentially all of the other jars in our application.  Here is the total list:
[jar-list.txt](https://github.com/lukehutch/fast-classpath-scanner/files/1026591/jar-list.txt)

We are using the scanner roughly like this:
```java
private static final String JAR_FILTER_REGEX = System.getProperty("zap.annotation.scanner.jar.regex", ".*");

private static final FastClasspathScanner CLASSPATH_SCANNER;

static {
      Set<String> scanSpec = new HashSet<>();

      FastClasspathScanner classpathFinder = new FastClasspathScanner();
      classpathFinder.getUniqueClasspathElements().stream()
         .map(File::getName)
         .filter(e -> e.matches(JAR_FILTER_REGEX))
         .forEach(e -> scanSpec.add("jar:" + e));

      CLASSPATH_SCANNER = new FastClasspathScanner(scanSpec.toArray(new String[0]));
}
...
public static Set<Class<?>> classesWithAnnotation(final Class<?> annotation) {
   final Set<Class<?>> matchingClasses = new HashSet<>();
   CLASSPATH_SCANNER.matchClassesWithAnnotation(annotation, matchingClasses::add).scan();

   return matchingClasses;
}
```
Note that we *do* define the ``zap.annotation.scanner.jar.regex`` System property, and it looks like this: ``zap.annotation.scanner.jar.regex=^com-dancer-.*`` ... which *should* restrict the scanning to the jars in your "jars-scanned-by-both" list.

So, now it seems to me that v2.0.10 was honoring the whitelist defined by that regex, whereas v2.0.11 does not seem to be.

**Comment by brettwooldridge** [comment]
@lukehutch I've tried to make things more explicit, to reduce scanning, but I'm not having much luck so maybe I'm missing something.

Again, my startup looks like this:
```java
Set<String> scanSpec = new LinkedHashSet<>();

FastClasspathScanner classpathFinder = new FastClasspathScanner();
classpathFinder.getUniqueClasspathElements()
   .stream()
   .map(File::getName)
   .forEach(e -> {
      if (e.matches("^com-dancer-.*")) {
         scanSpec.add("jar:" + e);
      }
      else {
         scanSpec.add("-jar:" + e);
      }
   });
```
This basically builds up a scanSpec collection with all of our jars included as ``jar:yada..`` and all third-party jars as ``-jar:yodel..``

Then we scan:
```java
CLASSPATH_SCANNER = new FastClasspathScanner(scanSpec.toArray(new String[0]))
   .verbose()
   .scan();
```
At first it looks good...

*Whitelisted jars:*
```
FastClasspathScanner	-- Whitelisted jars:  [com-dancer-bridge-events.jar, com-dancer-server-lucene.jar, com-dancer-bridge-telemetry.jar, com-dancer-provider-mibs-kt.jar, com-dancer-bridge-support.jar, com-dancer-perl.jar, com-dancer-service-hollow-kt.jar, com-dancer-zap-preference.jar, com-dancer-provider-adapters.jar, com-dancer-bridge-jobs.jar, com-dancer-bridge-backup.jar, com-dancer-bridge-update.jar, com-dancer-server-preferences.jar, com-dancer-zap-security.jar, com-dancer-net-connectivity.jar, com-dancer-bridge-collection.jar, com-dancer-server-core.jar, com-dancer-zap-jta.jar, com-dancer-bridge-jumphost.jar, com-dancer-derby.jar, com-dancer-provider-netman.jar, com-dancer-zap-migration.jar, com-dancer-server-reports.jar, com-dancer-provider-preferences.jar, com-dancer-server-cache.jar, com-dancer-provider-zerotouch.jar, com-dancer-provider-configstore.jar, com-dancer-server-help.jar, com-dancer-server-mail.jar, com-dancer-bridge-http.jar, com-dancer-provider-memos.jar, com-dancer-provider-update.jar, com-dancer-server-security.jar, com-dancer-tools.jar, com-dancer-server-redundancy.jar, com-dancer-server-corejobs.jar, com-dancer-bridge-discovery.jar, com-dancer-provider-devices.jar, com-dancer-provider-compliance.jar, com-dancer-net-snmp.jar, com-dancer-zap-util.jar, com-dancer-bridge-tools.jar, com-dancer-server-tools.jar, com-dancer-provider-backup.jar, com-dancer-provider-termlogs.jar, com-dancer-server-restore.jar, com-dancer-adapters.jar, com-dancer-reports-inventory.jar, com-dancer-server-discovery.jar, com-dancer-provider-jumphost.jar, com-dancer-bridge-proxy.jar, com-dancer-bridge-server.jar, com-dancer-provider-collection-hollow.jar, com-dancer-server-sshd.jar, com-dancer-provider-collection-kt.jar, com-dancer-bridge-hollow.jar, com-dancer-provider-tools.jar, com-dancer-net-adapters.jar, com-dancer-net-discovery.jar, com-dancer-net-servers.jar, com-dancer-server-dispatcher.jar, com-dancer-server-migration.jar, com-dancer-provider-dashboard-kt.jar, com-dancer-zap-util-kt.jar, com-dancer-zap-blazeds.jar, com-dancer-zap-bootstrap.jar, com-dancer-net-ftp.jar, com-dancer-provider-security.jar, com-dancer-server-bridge.jar, com-dancer-net-syslog.jar, com-dancer-provider-bulkupdate.jar, com-dancer-provider-credentials.jar, com-dancer-zap-rest.jar, com-dancer-provider-bridge.jar, com-dancer-provider-scheduler.jar, com-dancer-server-jobsdb.jar, com-dancer-server-web.jar, com-dancer-bridge-dns.jar, com-dancer-zap-web.jar, com-dancer-provider-events.jar, com-dancer-provider-telemetry.jar, com-dancer-net.jar, com-dancer-bridge-credentials.jar, com-dancer-provider-launchers.jar, com-dancer-server-sec.jar]
```
*Blacklisted jars:*
```
FastClasspathScanner	-- Blacklisted jars:  [hollow-2.5.2.jar, snakeyaml-1.17.jar, jetty-servlets-9.3.13.v20161014.jar, converter-moshi-2.1.0.jar, jasperreports-6.2.2.jar, json-io-2.5.1.jar, snmp4j-1.11.3.jar, javax.websocket-api-1.0.jar, jcifs-1.3.18.2.jar, slf4j-log4j12-1.7.21.jar, sshd-core-0.11.0.jar, derbytools-10.12.1.1.jar, log4j-api-2.8.1.jar, reflectasm-1.10.1.jar, javax.servlet-api-3.1.0.jar, curvesapi-1.04.jar, javax.annotation-api-1.2.jar, jetty-security-9.3.13.v20161014.jar, websocket-client-9.3.13.v20161014.jar, objenesis-2.1.jar, ecj-4.3.1.jar, xalan-2.6.0.jar, com.ibm.icu-52.1.0.v201404241930.jar, cron4j-2.2.5.jar, jta-2.6.jar, oshi-core-3.2.jar, appup-0.1.8.jar, commons-lang3-3.4.jar, jcommon-1.0.23.jar, jetty-rewrite-9.3.13.v20161014.jar, lucene-queries-6.2.1.jar, logging-interceptor-3.5.0.jar, base64-2.3.9.jar, jackson-core-2.7.2.jar, javax-websocket-client-impl-9.3.13.v20161014.jar, activation-1.1.jar, lucene-core-6.2.1.jar, javax.mail-api-1.5.6.jar, api-ldap-model-1.0.0-RC1.jar, jackson-module-jaxb-annotations-2.8.8.jar, jetty-webapp-9.3.13.v20161014.jar, jetty-http-9.3.13.v20161014.jar, lucene-backward-codecs-6.2.1.jar, jetty-util-9.3.13.v20161014.jar, log4j-jul-2.8.1.jar, portlet-api-2.0.jar, api-asn1-ber-1.0.0-RC1.jar, slf4j-api-1.7.21.jar, java-util-1.9.0.jar, okio-1.11.0.jar, derby-10.12.1.1.jar, kotlin-reflect-1.1.2.jar, ftplet-api-1.0.6.jar, gson-2.8.0.jar, jsmiparser-api-0.13.jar, okhttp-3.5.0.jar, asm-tree-5.0.1.jar, dhcp4java-1.0.jar, dnsjnio-1.0.3.jar, jetty-client-9.3.13.v20161014.jar, kotlin-stdlib-jre7-1.1.2.jar, commons-compress-1.12.jar, lucene-analyzers-common-6.2.1.jar, velocity-1.7.jar, jetty-servlet-9.3.13.v20161014.jar, jna-4.2.2.jar, websocket-servlet-9.3.13.v20161014.jar, kryo-3.0.3.jar, hollow-ui-tools-2.5.2.jar, commons-io-2.5.jar, oro-2.0.8.jar, poi-3.15.jar, commons-lang-2.6.jar, guava-20.0.jar, jackson-databind-2.7.2.jar, log4j-1.2-api-2.8.1.jar, poi-ooxml-schemas-3.15.jar, api-util-1.0.0-RC1.jar, javax.mail-1.5.6.jar, javax.transaction-api-1.2.jar, poi-ooxml-3.15.jar, websocket-common-9.3.13.v20161014.jar, commons-math3-3.6.1.jar, influxdb-java-2.5.jar, hollow-explorer-ui-2.5.2.jar, kerberos-client-patch-2.0.0-SNAPSHOT.jar, opencsv-2.4.jar, fast-classpath-scanner-2.0.21.jar, jetty-jndi-9.3.13.v20161014.jar, lucene-queryparser-6.2.1.jar, commons-digester-2.1.jar, jackson-dataformat-yaml-2.8.8.jar, commons-logging-1.2.jar, ftpserver-core-1.0.6.jar, mina-core-2.0.4.jar, minlog-1.3.0.jar, jetty-annotations-9.3.13.v20161014.jar, websocket-api-9.3.13.v20161014.jar, xercesImpl-2.11.0.jar, commons-beanutils-1.8.3.jar, api-i18n-1.0.0-RC1.jar, derbynet-10.12.1.1.jar, syslog4j-0.9.57.jar, jetty-util-ajax-9.3.13.v20161014.jar, xml-apis-1.4.01.jar, websocket-server-9.3.13.v20161014.jar, commons-net-3.5.jar, jline-1.0.jar, dnsjava-2.1.7.jar, flex-messaging-common-4.7.2.jar, flex-messaging-remoting-4.7.2.jar, funktionale-utils-1.0.1.jar, jetty-continuation-9.3.13.v20161014.jar, org.suigeneris.jrcs.diff-0.4.2.jar, commons-collections-3.2.1.jar, fst-2.48.jar, retrofit-2.1.0.jar, jetty-xml-9.3.13.v20161014.jar, jetty-jaas-9.3.13.v20161014.jar, lucene-highlighter-6.2.1.jar, lucene-memory-6.2.1.jar, woodstox-core-asl-4.4.1.jar, ehcache-2.8.5.jar, javax.persistence-2.1.1.jar, jetty-server-9.3.13.v20161014.jar, jfreechart-1.0.19.jar, kotlin-stdlib-1.1.2.jar, api-asn1-api-1.0.0-RC1.jar, moshi-1.2.0.jar, lucene-sandbox-6.2.1.jar, commons-codec-1.10.jar, annotations-13.0.jar, netty-all-4.0.42.Final.jar, derbyclient-10.12.1.1.jar, funktionale-option-1.0.1.jar, core-2.3.0.jar, postgresql-9.4.1212.jar, funktionale-collections-1.0.1.jar, kotlin-stdlib-jre8-1.1.2.jar, hollow-diff-ui-2.5.2.jar, jackson-module-kotlin-2.8.8.jar, itext-2.1.7.js5.jar, javax.ejb-api-3.2.jar, joda-time-2.9.4.jar, nuprocess-1.1.0.jar, jetty-io-9.3.13.v20161014.jar, tinyradius-1.0.jar, apacheds-kerberos-codec-2.0.0-M23.jar, asm-5.0.1.jar, apacheds-i18n-2.0.0-M23.jar, javassist-3.19.0-GA.jar, jsmiparser-util-0.13.jar, sakai-rwiki-diff-10.3.jar, funktionale-either-1.0.1.jar, stax2-api-3.1.4.jar, log4j-core-2.8.1.jar, javax-websocket-server-impl-9.3.13.v20161014.jar, flex-messaging-core-4.7.2.jar, asm-commons-5.0.1.jar, dynamicreports-core-4.1.1.jar, jetty-plus-9.3.13.v20161014.jar, lucene-join-6.2.1.jar, jackson-annotations-2.8.8.jar, jsonrpc4j-1.4.6.jar, jna-platform-4.2.2.jar, antlr-2.7.7.jar, HikariCP-2.6.0.jar, jetty-blazeds-8.0.4.v20111024.jar, sansorm-2.4.jar, commons-collections4-4.1.jar]
```

Then things start looking not so bright... but at first this:
```
FastClasspathScanner	-- Finding classpath elements in ClassLoader sun.misc.Launcher$AppClassLoader@18b4aac2
FastClasspathScanner	---- Adding classpath element: file:/Users/brettw/Documents/lvi/thirdeye/dist/server/core/activation-1.1.jar
...
FastClasspathScanner	---- Adding classpath element: file:/Users/brettw/Documents/lvi/thirdeye/dist/server/core/xml-apis-1.4.01.jar
```
So it finds *all* the jars in the AppClassLoader without regard to white/black list ...

And then...
```
FastClasspathScanner	-- Getting classpath entries from java.class.path
FastClasspathScanner	---- Adding classpath element: core/activation-1.1.jar
...
FastClasspathScanner	---- Adding classpath element: core/xml-apis-1.4.01.jar
```
It adds all the same jars again (without respect to white/black list) only this time with a *relative path* instead of absolute.

And finally...
```
FastClasspathScanner	-- Scanning jarfile /Users/brettw/Documents/lvi/thirdeye/dist/server/core/annotations-13.0.jar (took 0.004641 sec)
FastClasspathScanner	---- Found whitelisted file in jarfile: META-INF/MANIFEST.MF
FastClasspathScanner	---- Found whitelisted file in jarfile: org/intellij/lang/annotations/Flow.class
```
Proceeds to discover all the files in all the jars (that are not only not whitelisted but are specifically blacklisted).

After a bunch of that...
```
Ignoring duplicate classpath element: /Users/brettw/Documents/lvi/thirdeye/dist/server/core/activation-1.1.jar
...
FastClasspathScanner	-- Ignoring duplicate classpath element: /Users/brettw/Documents/lvi/thirdeye/dist/server/core/xml-apis-1.4.01.jar
```
and
```
FastClasspathScanner	Ignoring jarfile, because it is not whitelisted: /Users/brettw/Documents/lvi/thirdeye/dist/server/core/activation-1.1.jar
...
FastClasspathScanner	Ignoring jarfile, because it is not whitelisted: /Users/brettw/Documents/lvi/thirdeye/dist/server/core/xml-apis-1.4.01.jar
```
Why is whitelisting and blacklisting not applied earlier?

**Comment by lukehutch** [comment]
Apologies, I'm on a road trip in Norway for the next few days, will be slow to respond. However, paths are normalized (the raw URLs or paths obtained from each ClassLoader are displayed, but they are normalized before being scanned, and duplicates are removed, so getting the same URLs twice from two different ClassLoaders, with relative and absolute paths, shouldn't matter). (In more recent versions of FastClasspathScanner, typically only one ClassLoader is consulted for URLs, reducing duplication.) All the paths are obtained from ClassLoaders then filtered for whitelist/blacklist patterns -- filtering earlier wouldn't speed things up though, and the time taken for finding and filtering paths is negligible. The main thing you need to fix is the fact you scan 13x rather than once. I'll look at the other things in your comments more closely once I get back to civilization.

**Comment by brettwooldridge** [comment]
@lukehutch Thanks, we've already fixed our code to scan once.  I'm more concerned about the scanner parsing the class info for unnecessary classes from jars outside of the whitelist:
```
FastClasspathScanner	---- Found whitelisted file in jarfile: org/intellij/lang/annotations/Flow.class
```

**Comment by lukehutch** [comment]
@brettwooldridge OK, I figured out what was going on. This took some thinking through, because the issue that caused the slowdown is very complex: #100. Basically, the issue is that FastClasspathScanner should always follow the principle of minimum surprise. If you have the same class `X` defined in jars `A.jar` and `B.jar`, and the classpath order is `A.jar:B.jar`, and your whitelist contains only `B.jar`, then if you try loading the class `X`, it will load from `A.jar`, since that is the first definition that the current classloader will find. But FastClasspathScanner will give you info from the definition of `X` found in `B.jar`, which may not agree with the version of the class loaded by the system. (Even more complex is the case where you use a `MatchProcessor`, which will cause FastClasspathScanner to potentially load the version from `B.jar`, or it may reuse a cached version already loaded by the system from `A.jar` if the class has already been referenced -- the behavior is for all intents and purposes undefined).

The slowdown you discovered is due to needing to scan all the files in the entire classpath, whether or not a jar is whitelisted, so that the definition of `X` in `A.jar` can mask the definition of the same class in `B.jar`, so that if FastClasspathScanner is given just `B.jar` as a whitelist, it should not see the definition of class `X` at all. This was an intentional choice, to prevent the wrong version of the class being loaded, which has been a recurring issue (at least three different bugs have been filed about this in different forms, so a lot of people must work with classpaths that contain multiple different definitions of the same class).

I don't know that the slowdown is worth the avoidance of confusion here, so I think I'm going to revert part of this behavior, and if there are multiple definitions of any given class, I think I'm just going to have to tell users that all bets are off as to which class will actually match or get loaded, if they also try to use whitelist/blacklist criteria.

Hopefully that all makes sense...

**Comment by lukehutch** [comment]
@brettwooldridge Please test the latest version in git. The code should be back to running full speed, the order of whitelisting has been fixed, and the logging has been improved.

**Comment by lukehutch** [comment]
@brettwooldridge I will be pushing out a new release in the next couple of days. Please let me know if this issue is fixed for you using the latest git master version. Thank you!

**Comment by brettwooldridge** [comment]
@lukehutch Thanks a lot, man! 👍  Excellent project!

**Comment by lukehutch** [comment]
Thanks @brettwooldridge -- although can you please test and let me know if the problem is fixed in git before I push out a release?

**Comment by brettwooldridge** [comment]
@lukehutch I just tested 2.0.22-SNAPSHOT.  Our startup times are back down into the range they previously were.

**Comment by lukehutch** [comment]
@brettwooldridge great, thanks for testing. To be doubly emphatic about it, be aware that if you use jar whitelisting/blacklisting or classpath overriding, you may end up with jars scanned in a different order than the system classloader opens them to load classes, so you will get unpredictable results if a class is defined more than once in the classpath. (This situation occurs more often than you would think.)

**Comment by lukehutch** [comment]
The fix was released in 2.0.22. Thanks!
